java/meterpreter/meterpreter/src/main/java/com/metasploit/meterpreter/Transport.java

```diff
@@ -45,6 +47,60 @@ protected int parseTimeouts(byte[] configuration, int offset) {
         return offset;
     }
 
+    protected TLVPacket readAndDecodePacket(DataInputStream in) throws IOException {
+        int xorKey = in.readInt();
+        int len = (in.readInt() ^ Integer.reverseBytes(xorKey)) - 8;
+        int type = in.readInt();
+        byte[] body = new byte[len];
+        in.readFully(body);
+
+        this.xorBytes(xorKey, body);
+
+        ByteArrayInputStream byteStream = new ByteArrayInputStream(body);
+        DataInputStream inputStream = new DataInputStream(byteStream);
+        TLVPacket packet = new TLVPacket(inputStream, len);
+        inputStream.close();
+
+        return packet;
+    }
+
+    protected void encodePacketAndWrite(TLVPacket packet, int type, DataOutputStream out) throws IOException {
+        int xorKey = randXorKey();
+        byte[] data = packet.toByteArray();
+        this.xorBytes(xorKey, data);
+        synchronized (out) {
+            out.writeInt(xorKey);
+            out.writeInt((data.length + 8) ^ Integer.reverseBytes(xorKey));
```

Comment:
Yes, we lifted to 1.5 a few months back, largely to to assist in Android development.